opal/util/opal_gethostname.c

```diff
@@ -0,0 +1,178 @@
+                 * could get caught in the loop. So, break after
+                 * NUM_TRIES_FOR_NULL_HOSTNAME. 
+                 */
+                else if( 0 == count ) {
+                    if( num_tries > NUM_TRIES_FOR_NULL_HOSTNAME ) {
+                        ret_val = -1;
+                        temp_errno = EINVAL;
+                        break;
+                    }
+                    num_tries++;
+                }
+                /* The result *may* be truncated (can't know for sure, so we increase 
+                 * the length of the buffer (happens below) and try again). */
+                else if( count == length - 1 ) {
+			; /* no op */
+                }
+                /* Unidentified case, implies some bug in this code or some unknown
+                 * return string from gethostname() */
+                else {
+                    strcpy( buf, UNKNOWN_ERROR_STRING );
+                    break;
+                }
+            }
+            /*
+             * hostname was truncated and there was an error. Perhaps
+             * there is something in the buffer and perhaps not. 
+             * Set temp_errno to the most appropriate value;
+             * Double the length and try again. 
+             */
+            else if( EINVAL == errno || ENAMETOOLONG == errno ) {
+                temp_errno = ENAMETOOLONG;
+            }
+            /*
+             * Memory error (invalid address). Have we run out of memory?
+             */
+            else if( EFAULT == errno ) {
+                temp_errno = errno;
+                strcpy( buf, MEMORY_ERROR_STRING );
+                break;
+            }
+            /* Unidentified case, implies some unknown error returned from gethostname() */
+            else {
+                temp_errno = errno;
+                strcpy( buf, UNKNOWN_ERROR_STRING );
+                break;
+            }
+            length *= 2;
+            buf = realloc( buf, length );
+            if( NULL == buf ) {
+               errno = ENOMEM;
+               return MALLOC_ERROR_STRING;
+            }
+        }
+        /*
+         * Eventually, we would like this code to allow the full string
+         * to be copied in all (or all reasonable) cases (arbitrary length)
+         */
+        strncpy( opal_local_hostname, buf, OPAL_LOCAL_MAXHOSTNAMELEN - 1 );
+        opal_local_hostname[OPAL_LOCAL_MAXHOSTNAMELEN - 1] = '\0';
+        free( buf );
+        /* restore errno if there was a prior error to preserve error condition */
+        if( 0 != ret_val ) {
+            errno = temp_errno;
+        }
+        /* if there was not an error, but the result is truncated, 
+         * we set errno appropriately */
+        else if( length > OPAL_LOCAL_MAXHOSTNAMELEN ) {
+            errno = ENAMETOOLONG;
+        }
+    }
+
+    /* Re-copy hostname into a temp variable because of risk
+       of caller modifying this variable. */
+    strcpy(temp_hostname, opal_local_hostname);
```

Comment:
Fair point that some other caller may have modified the buffer.  But since we're returning a `const char *`, is that actually a possibility?

Reply:
Yes, it can still happen under certain conditions - this is in the conversation about orte/util/proc_info.c here:
https://github.com/open-mpi/ompi/pull/7200#issuecomment-566186017

Reply:
Ah, gotcha!

Given that you've now had to explain it twice 😄, that means it's probably worth describing in a comment so that our future selves don't have this conversation again...